src/main/java/net/neoforged/neoforge/common/crafting/outgredient/Outgredient.java

```diff
@@ -0,0 +1,130 @@
+        return new DefaultedItemTagOutgredient(tagKey, fallback.asItem().builtInRegistryHolder());
+    }
+
+    /**
+     * Creates a new {@link TagKey}-backed {@link Outgredient} without a fallback.
+     *
+     * @param tagKey The {@link TagKey} backing the {@link Outgredient}.
+     * @return A new {@link Outgredient}.
+     */
+    static Outgredient<ItemStack> ofItem(TagKey<Item> tagKey) {
+        return new DefaultedItemTagOutgredient(tagKey, Optional.empty());
+    }
+
+    /**
+     * Creates a new {@link TagKey}-backed {@link Outgredient} with an {@link ItemLike} fallback.
+     *
+     * @param tagKey   The {@link TagKey} backing the {@link Outgredient}.
+     * @param fallback The fallback {@link ItemLike}.
+     * @param count    The output count.
+     * @return A new {@link Outgredient}.
+     */
+    static Outgredient<ItemStack> ofItem(TagKey<Item> tagKey, ItemLike fallback, int count) {
+        return new DefaultedItemTagOutgredient(tagKey, fallback.asItem().builtInRegistryHolder());
+    }
+
+    /**
+     * Creates a new {@link TagKey}-backed {@link Outgredient} without a fallback.
+     *
+     * @param tagKey The {@link TagKey} backing the {@link Outgredient}.
+     * @param count  The output count.
+     * @return A new {@link Outgredient}.
+     */
+    static Outgredient<ItemStack> ofItem(TagKey<Item> tagKey, int count) {
+        return new DefaultedItemTagOutgredient(tagKey, Optional.empty());
+    }
+
+    /**
+     * Creates a new {@link FluidStack}-backed {@link Outgredient}.
+     *
+     * @param stack The {@link FluidStack} backing the {@link Outgredient}.
+     * @return A new {@link Outgredient}.
+     */
+    static Outgredient<FluidStack> ofFluid(FluidStack stack) {
+        return new OutgredientWrapper<>(stack, new FluidStackSlotDisplay(stack));
+    }
+
+    /**
+     * Creates a new {@link TagKey}-backed {@link Outgredient} with an {@link Fluid} fallback.
+     *
+     * @param tagKey   The {@link TagKey} backing the {@link Outgredient}.
+     * @param fallback The fallback {@link Fluid}.
+     * @param amount   The output amount.
+     * @return A new {@link Outgredient}.
+     */
+    static Outgredient<FluidStack> ofFluid(TagKey<Fluid> tagKey, Fluid fallback, int amount) {
+        return new DefaultedFluidTagOutgredient(tagKey, fallback.builtInRegistryHolder(), amount);
+    }
+
+    /**
+     * Creates a new {@link TagKey}-backed {@link Outgredient} without a fallback.
+     *
+     * @param tagKey The {@link TagKey} backing the {@link Outgredient}.
+     * @param amount The output amount.
+     * @return A new {@link Outgredient}.
+     */
+    static Outgredient<FluidStack> ofFluid(TagKey<Fluid> tagKey, int amount) {
+        return new DefaultedFluidTagOutgredient(tagKey, Optional.empty(), amount);
+    }
+
+    /**
+     * Resolves the outgredient into a specific {@code T}.
+     *
+     * @return The resolved outgredient.
+     */
+    T resolve();
```

Comment:
This very likely needs a `RegistryAccess` parameter like recipes have. Will let the result depend on a datapack registry value that is not cached in the object.